### PR TITLE
Add PHP 8.2 support

### PR DESCRIPTION
**Pull request by fezfez**
Signed-off-by: Stéphane Demonchaux <demonchaux.stephane@gmail.com>

### CODE REVIEW

**Comment by driehle** [comment]
Looks like `phpspec/prophecy` doesn't support 8.2 yet...

**Comment by driehle** [comment]
I dropped `phpspec/prophecy` with #65. Could you please rebase your PR on the latest `3.2.x` branch?

**Comment by fezfez** [comment]
> I dropped `phpspec/prophecy` with #65. Could you please rebase your PR on the latest `3.2.x` branch?

done

**Comment by fezfez** [comment]
@driehle seem to be good :)

**Comment by driehle** [comment]
Looks good, but I a wondering if we should add a test case for readonly properties (PHP 8.1) and readonly classes (PHP 8.2). As uninitialized properties are ignored for extraction by reference...

https://github.com/doctrine/doctrine-laminas-hydrator/blob/3.2.x/src/DoctrineObject.php#L288

... readonly properties could be ignored for hydration by reference. What do you think?

**Comment by fezfez** [comment]
@driehle tests has beed added :)

File: tests/DoctrineObjectTest.php [hunk 1]
```diff
@@ -2978,4 +2979,49 @@ public function testWrongEnumBackedValueThrowsException(): void
...
+    }
+
+    /**
+     * @requires PHP 8.1
+     */
+    public function testHydrateReadonlyProps(): void
+    {
+        $this->configureObjectManagerForSimpleEntity(Assets\SimpleEntityWithReadonlyPropsPhp81::class);
+
+        $entity = $this->hydratorByValue->hydrate(['field' => 'toto', 'id' => 15], new Assets\SimpleEntityWithReadonlyPropsPhp81(3));
```
Comment (driehle) [inline tests/DoctrineObjectTest.php:3026]:
```suggestion
        $entity = $this->hydratorByReference->hydrate(['field' => 'toto', 'id' => 15], new Assets\SimpleEntityWithReadonlyPropsPhp81(3));
```

This test should hydrate by reference. Hydrating by value, i.e. by calling getters/setters, will be fine, as long as the user doesn't add setters for a readonly property - which then would clearly be a user error. The case we should test here is that our library should not try to modify properties via reflection if they are readonly.

File: tests/DoctrineObjectTest.php [hunk 2]
```diff
@@ -2978,4 +2979,49 @@ public function testWrongEnumBackedValueThrowsException(): void
...
+
+        $this->assertEquals(3, $entity->getId());
+        $this->assertEquals('toto', $entity->getField());
+    }
+
+    /**
+     * @requires PHP 8.2
+     */
+    public function testExtractReadonlyClass(): void
```
Comment (driehle) [inline tests/DoctrineObjectTest.php]: Could you please add a test `testHydrateReadonlyClass()` as well?

For this, you will need to add a check and throw of `LogicException`  in `DoctrineObject::hydrateByReference()`.

File: src/DoctrineObject.php [hunk 3]
```diff
@@ -277,6 +277,16 @@ protected function extractByReference(object $object): array
             : $this->filterComposite;
 
         $data = [];
+
+        if (method_exists($refl, 'isReadOnly') && $refl->isReadOnly()) {
```
Comment (driehle) [inline src/DoctrineObject.php:282]:
```suggestion
        // fail for readonly classes (PHP 8.1)
        if (method_exists($refl, 'isReadOnly') && $refl->isReadOnly()) {
```

**Review by driehle — CHANGES_REQUESTED** [review]

File: src/DoctrineObject.php (hunk 3 above)
Comment (fezfez) [reply src/DoctrineObject.php:282]: done

File: tests/DoctrineObjectTest.php (hunk 2 above)
Comment (fezfez) [reply tests/DoctrineObjectTest.php]: done

File: tests/DoctrineObjectTest.php (hunk 1 above)
Comment (fezfez) [reply tests/DoctrineObjectTest.php:3026]: done

**Comment by fezfez** [comment]
@driehle  should be ok

**Review by driehle — APPROVED** [review]

**Comment by driehle** [comment]
Thanks, @fezfez, for you time and contribution!
3.2.0 has just been released!

**Pull request merged**